app/scripts/controller-init/types.ts

```diff
@@ -0,0 +1,168 @@
+  ActionConstraint,
+  EventConstraint
+>;
+
+/** Generic restricted controller messenger using base template types. */
+export type BaseRestrictedControllerMessenger = RestrictedControllerMessenger<
+  string,
+  ActionConstraint,
+  EventConstraint,
+  string,
+  string
+>;
+
+/**
+ * Request to initialize and return a controller instance.
+ * Includes standard data and methods not coupled to any specific controller.
+ */
+export type ControllerInitRequest<
+  ControllerMessengerType extends void | BaseRestrictedControllerMessenger = void,
+  InitMessengerType extends void | BaseRestrictedControllerMessenger = void,
+> = {
+  /**
+   * Required controller messenger instance.
+   * Generated using the callback specified in `getControllerMessengerCallback`.
+   */
+  controllerMessenger: ControllerMessengerType extends BaseRestrictedControllerMessenger
+    ? ControllerMessengerType
+    : never;
+
+  /**
+   * Retrieve a controller instance by name.
+   * Throws an error if the controller is not yet initialized.
+   *
+   * @param name - The name of the controller to retrieve.
+   */
+  getController<Name extends ControllerName>(
+    name: Name,
+  ): ControllerByName[Name];
+
+  /**
+   * Retrieve the flat state for all controllers.
+   * For example: `{ transactions: [] }`.
+   *
+   * @deprecated Subscribe to other controller state via the messenger.
+   */
+  getFlatState: () => ControllerFlatState;
+
+  /**
+   * Retrieve the chain ID of the globally selected network.
+   *
+   * @deprecated Will be removed in the future pending multi-chain support.
+   */
+  getGlobalChainId(): string;
+
+  /**
+   * Retrieve the permitted accounts for a given origin.
+   *
+   * @param origin - The origin for which to retrieve permitted accounts.
+   * @param options - Additional options for the request.
+   * @param options.suppressUnauthorizedError - Whether to not throw if an unauthorized error occurs. Defaults to `true`.
+   */
+  getPermittedAccounts(
+    origin: string,
+    options?: { suppressUnauthorizedError?: boolean },
+  ): Promise<string[]>;
+
+  /**
+   * Retrieve the provider instance for the globally selected network.
+   *
+   * @deprecated Will be removed in the future pending multi-chain support.
+   */
+  getProvider: () => Provider;
+
+  /**
+   * Retrieve a transaction metrics request instance.
+   * Includes data and callbacks required to generate metrics.
+   */
+  getTransactionMetricsRequest(): TransactionMetricsRequest;
+
+  /**
+   * Required initialization messenger instance.
+   * Generated using the callback specified in `getInitMessengerCallback`
+   */
+  initMessenger: InitMessengerType extends BaseRestrictedControllerMessenger
+    ? InitMessengerType
+    : never;
+
+  /**
+   * The full persisted state for all controllers.
+   * Includes controller name properties.
+   * e.g. `{ TransactionController: { transactions: [] } }`.
+   */
+  persistedState: ControllerPersistedState;
```

Comment:
Adding `Partial` here handles controllers that don't get persisted, and it handles the case where a user is missing persisted state for one or more controllers (because they're updating from an older version that lacked them).
```suggestion
  persistedState: Partial<ControllerPersistedState>;
```

Reply:
Or maybe we could include the wrapper `Partial` as part of `ControllerPersistedState`? Maybe that would make more sense semantically.